py/kubeflow/testing/run_e2e_workflow.py

```diff
@@ -89,9 +101,13 @@ def parse_config_file(config_file, root_dir):
 
   components = []
   for i in results["workflows"]:
-    components.append(WorkflowComponent(
-      i["name"], os.path.join(root_dir, i["app_dir"]), i["component"], i.get("job_types", []),
-      i.get("include_dirs", []), i.get("params", {})))
+    if i.get("app_dir"):
+      components.append(WorkflowKSComponent(
+        i["name"], os.path.join(root_dir, i["app_dir"]), i["component"],
+        i.get("job_types", []), i.get("include_dirs", []), i.get("params", {})))
+    if i.get("py_func"):
+      components.append(WorkflowPyComponent(
+        i["py_func"], i.get("kw_args", {})))
   return components
 
 def generate_env_from_head(args):
```

Comment:
On line 200
```
    workflow_name = os.getenv("JOB_NAME") + "-" + w.name
    ks_cmd = ks_util.get_ksonnet_cmd(w.app_dir)
```
should move to under line 259
```
    if w.app_dir:
```

Reply:
Can you explain this? `workflow_name` is being used between line 201 and line 252.

Reply:
either py_func needs a name as well or we use py_func in place of w.name

Reply:
Ah, I think I'm understanding what you're saying, ha.

Reply:
Fixed.

Reply:
Sigh, I thought I fixed it. I don't get why it's still throwing an error since it shouldn't include `WorkflowPyComponent` given that the workflow doesn't have an `app_dir` property...

```
  File "/src/kubeflow/testing/py/kubeflow/testing/run_e2e_workflow.py", line 196, in run
    if w.app_dir:
AttributeError: 'WorkflowPyComponent' object has no attribute 'app_dir'
```

Reply:
It might be worth updating the unittest so we verify part of the code is working just by running a unittest
https://github.com/kubeflow/testing/blob/master/py/kubeflow/tests/run_e2e_workflow_test.py

Refactoring the code may help.

In particular; IIUC the part of code that is causing problems is converting a prow_config.yaml file into a set of workflows and then verifying in the case of PyFunc we can extract a dictionary containing the expected Argo workflow.

Can we validate that using a unittest?

Reply:
Expanding on that a bit. I might suggest narrowing the scope of this first PR to just implement the following function

```
def build_workflows(prow_config):
   """
       Returns: 
           workflows: List of dictionarys containing Argo workflows
    """
    ....
```

* This function would take a dictionary containing the contents of the prow_config.yaml file and return a list of dictionaries containing argo workflows
* This function should only handle py_func workflows; entries defining ksonnet based workflows should just be skipped

In follow on PRs we can integrate this into the overall workflow. I think we can do that by refactoring the code to something like the following

```
def run(...)
  ... various setup ...
   prow_config = load prow_config
   ks_workflows = submit_ksonnet_workflows(prow_config)
   py_workflows = build_workflows(prow_config)
   submit_argo_workflows(py_workflows)

   all_workflows = ks_workflows + py_workflows
   wait_for_all_workflows(all_workflows)
```\
* So basically we would refactor the ksonnet specific bits of run into submit_ksonnet_workflows